### PR TITLE
339 add mypy static type checker

### PR DESCRIPTION
**Pull request by hokiegeek2**
This draft PR corresponds to add mypy static type checking, which is the complement to the runtime checking implemented in #540. 

There's a few more checks to implement along with CI/CD integration, both of which I am working on today

### CODE REVIEW

File: arkouda/decorators.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
+from typing import Union
+from arkouda import pdarray
+from arkouda.strings import Strings
+
+def checkforpdarray(func):
+    def check(pda : pdarray, **args):
+        if not isinstance(pda, pdarray):
+            raise TypeError('must be a pdarray, not a {}'.\
+                        format(pda.__class__.__name__))
+        return func(pda)
```
Comment (reuster986) [inline arkouda/decorators.py:10]: I'm new to decorators, but do `args` need to be passed to `func` somehow?

Comment (hokiegeek2) [reply arkouda/decorators.py:10]: Good point, yes, and it's kinda handled under the covers so to speak

**Comment by hokiegeek2** [comment]
This is odd: getting macos-specific test error that I can't replicate on my macbook:

image

image

File: arkouda/dtypes.py [hunk 2]
```diff
@@ -63,13 +63,13 @@ def translate_np_dtype(dt: np.dtype) -> Tuple[builtins.str, int]:
     kind = trans[dt.kind]
     return kind, dt.itemsize
 
-def resolve_scalar_dtype(val : object) -> str:
+def resolve_scalar_dtype(val : Union[np.bool,np.float64,int,np.int,np.uint8]) -> str: #type: ignore
```
Comment (reuster986) [inline arkouda/dtypes.py:66]: This function should accept all scalar types, not just numpy dtypes. Can we change it back to `object`?

File: arkouda/dtypes.py [hunk 3]
```diff
@@ -63,13 +63,13 @@ def translate_np_dtype(dt: np.dtype) -> Tuple[builtins.str, int]:
...
     """
     Try to infer what dtype arkouda_server should treat val as.
     """
     # Python bool or np.bool
     if isinstance(val, builtins.bool) or (hasattr(val, 'dtype') \
-                                and val.dtype.kind == 'b'):
+                                and cast(np,val).dtype.kind == 'b'):
```
Comment (reuster986) [inline arkouda/dtypes.py:72]: What does this cast do? I can't figure it out...

**Review by reuster986 — COMMENTED** [review]
I asked a couple questions in `dtypes.py` that I'd like to clarify before approving.

File: arkouda/dtypes.py (hunk 3 above)
Comment (hokiegeek2) [reply arkouda/dtypes.py:72]: @reuster986 bwoops, not much, will fix, thanks for catching that!

File: arkouda/dtypes.py (hunk 2 above)
Comment (hokiegeek2) [reply arkouda/dtypes.py:66]: @reuster986 Gotcha, absolutely, will do

File: arkouda/dtypes.py (hunk 3 above)
Comment (hokiegeek2) [reply arkouda/dtypes.py]: @reuster986 fixed

File: arkouda/dtypes.py (hunk 2 above)
Comment (hokiegeek2) [reply arkouda/dtypes.py]: @reuster986 fixed

File: arkouda/decorators.py (hunk 1 above)
Comment (hokiegeek2) [reply arkouda/decorators.py]: @reuster986 I moved this code out and put the type checking decorators in [PR 543](https://github.com/mhmerrill/arkouda/pull/543)

**Comment by hokiegeek2** [comment]
@reuster986 please take another look when you get a chance and let me know if there are any other changes you'd like to discuss.

**Review by reuster986 — APPROVED** [review]
Looks good to me.

**Comment by hokiegeek2** [comment]
@mhmerrill @reuster986 this PR is ready for final review now that I've added documentation and integrated mypy type checking into the make file as well as the CI workflow.

**Pull request merged**